pkg/sharings/send_mails.go

```diff
@@ -0,0 +1,222 @@
+// filled through the `mailTemplateValues` structure.
+const (
+	mailTemplateEn = `
+        <h2>Hey {{.RecipientName}}!</h2>
+        <p>{{.SharerPublicName}} wants to share a document with you! You will only be able to view it.</p>
+
+        <p>The description given is: {{.Description}}.</p>
+
+        <form action="{{.OAuthQueryString}}">
+            <input type="submit" value="See this document" />
+        </form>
+        </p>
+    `
+
+	mailTemplateFr = `
+        <h2>Bonjour {{.RecipientName}} !</h2>
+        <p>{{.SharerPublicName}} veut partager un document avec vous ! Vous pourrez seulement le consulter.</p>
+
+        <p>La description associée est : {{.Description}}.</p>
+
+        <form action="{{.OAuthQueryString}}">
+            <input type="submit" value="Voir le document" />
+        </form>
+    `
+)
+
+// The sharing-dependent information: the recipient's name, the sharer's public
+// name, the description of the sharing, and the OAuth query string.
+type mailTemplateValues struct {
+	RecipientName    string
+	SharerPublicName string
+	Description      string
+	OAuthQueryString string
+}
+
+// SendSharingMails will generate the mail containing the details
+// regarding this sharing, and will then send it to all the recipients.
+func SendSharingMails(instance *instance.Instance, s *Sharing) error {
+
+	// Generate the base values of the email to send, common to all recipients:
+	// the description and the sharer's public name.
+	mailValues := &mailTemplateValues{}
+
+	// We get the Couchdb document describing the instance to get the sharer's
+	// public name.
+	doc := &couchdb.JSONDoc{}
+	err := couchdb.GetDoc(instance, consts.Settings,
+		consts.InstanceSettingsID, doc)
+	if err != nil {
+		return err
+	}
+	sharerPublicName, _ := doc.M["public_name"].(string)
+	mailValues.SharerPublicName = sharerPublicName
+
+	// Fill in the description.
+	if s.Desc == "" {
+		mailValues.Description = "[No description provided]"
+	} else {
+		mailValues.Description = s.Desc
+	}
+
+	// For each recipient we send the email and update the status if an error
+	// occurred.
+	recipientsStatus, err := s.RecStatus(instance)
+	if err != nil {
+		return err
+	}
+
+	errorOccurred := false
+	for _, rs := range recipientsStatus {
+		recipient := rs.recipient
+
+		// Generate recipient specific OAuth query string.
+		recipientOAuthQueryString, err := generateOAuthQueryString(s, recipient)
+		if err != nil {
+			errorOccurred = logErrorAndSetRecipientStatus(rs, err)
+			break
+		}
+
+		// Augment base values with recipient specific information.
+		mailValues.RecipientName = recipient.Email
+		mailValues.OAuthQueryString = recipientOAuthQueryString
+
+		sharingMessage, err := generateMailMessage(s, recipient, mailValues)
+		if err != nil {
+			errorOccurred = logErrorAndSetRecipientStatus(rs, err)
+			break
+		}
+
+		// We ask to queue a new mail job.
+		// XXX Do we have to do something with the returned values?
```

Comment:
Not necessarily.

The first argument contains the initial job informations built synchronously and the second one is a channel on which you can listen for the advanced of the job.